File: 2020/355.py
"""
355
"""
import heapq
class Twitter:

    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.seenUser = {}          # key: id, val: list of newsFeed
        self.followingOfId = {}     # key: id, val: set of following KOLs
        self.t = 30009              # decreasing: easier for heap operation

    def _initUser(self, id):
        self.seenUser[id] = []
        self.followingOfId[id] = set([id])      # follow themselves

    def postTweet(self, userId: int, tweetId: int) -> None:
        """
        Compose a new tweet.
        """
        if userId not in self.seenUser: self._initUser(userId)
        self.seenUser[userId] += (self.t, tweetId),
        self.t -= 1
        return

    def getNewsFeed(self, userId: int) -> List[int]:
        """
        Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent.
        """
        res = []
        sources = []
        tmp = {}
        if userId not in self.seenUser: return []
        for kol in self.followingOfId[userId]:
            if len(self.seenUser[kol]) != 0:                # has some post
                tmp[kol] = len(self.seenUser[kol])-1        # get the latest post of kol
                t, tweetId = self.seenUser[kol][-1]
                heapq.heappush(sources, (t, tweetId, kol))
        
        while len(res) < 10 and sources:
            _, tid, kol = heapq.heappop(sources)
            tmp[kol] -= 1
            if tmp[kol] >= 0:
                t, newTid = self.seenUser[kol][tmp[kol]]
                heapq.heappush(sources, (t, newTid, kol))
            res.append(tid)

        return res
        

    def follow(self, followerId: int, followeeId: int) -> None:
        """
        Follower follows a followee. If the operation is invalid, it should be a no-op.
        """
        if followerId not in self.seenUser: self._initUser(followerId)
        if followeeId not in self.seenUser: self._initUser(followeeId)
        self.followingOfId[followerId].add(followeeId)
        return

    def unfollow(self, followerId: int, followeeId: int) -> None:
        """
        Follower unfollows a followee. If the operation is invalid, it should be a no-op.
        """
        if followerId not in self.seenUser: self._initUser(followerId)
        if followeeId not in self.seenUser: self._initUser(followeeId)
        if followeeId in self.followingOfId[followerId]: self.followingOfId[followerId].remove(followeeId)
        return


# Your Twitter object will be instantiated and called as such:
# obj = Twitter()
# obj.postTweet(userId,tweetId)
# param_2 = obj.getNewsFeed(userId)
# obj.follow(followerId,followeeId)
# obj.unfollow(followerId,followeeId)

File: current_session/c++/355.cpp
#include <iostream>
#include <vector>
#include <map>
#include <set>

using namespace std;

// usrpost map: key=user, val=list of tweets
// usrfollow map: key=user, val=set of following kols
// follow/unfollow: insert/remove from usrfollow[user]
// getNewsFeed: for all kols in usrfollow[user], get all posts from usrpost[kol]
// post: add into userPost map with timestamp
// sort and get top 10.

class Twitter {
public:
    map<int, vector<pair<int, int>>> usrPost;
    map<int, set<int>> usrFollow;
    int timestamp = 30003;
    /** Initialize your data structure here. */
    Twitter() {  
    }
    
    /** Compose a new tweet. */
    void postTweet(int userId, int tweetId) {
      if (usrPost.find(userId) == usrPost.end()) {usrFollow[userId].insert(userId);}
      usrPost[userId].push_back(pair<int, int>{--timestamp,tweetId});
    }
    
    /** Retrieve the 10 most recent tweet ids in the user's news feed. Each item in the news feed must be posted by users who the user followed or by the user herself. Tweets must be ordered from most recent to least recent. */
    vector<int> getNewsFeed(int userId) {
      vector<pair<int,int>> feed;
      for (auto kol: usrFollow[userId]) {
        for (auto t: usrPost[kol]) {
          feed.push_back(pair<int, int>{t.first, t.second});
        }
      }
      // sort
      sort(feed.begin(), feed.end());
      vector<int> res;
      for (int i=0; i<feed.size() && i<10; ++i) {
        res.push_back(feed[i].second);
      }
      return res;
    }
    
    /** Follower follows a followee. If the operation is invalid, it should be a no-op. */
    void follow(int followerId, int followeeId) {
      usrFollow[followerId].insert(followeeId);
    }
    
    /** Follower unfollows a followee. If the operation is invalid, it should be a no-op. */
    void unfollow(int followerId, int followeeId) {
      usrFollow[followerId].erase(followeeId);
    }
};

/**
 * Your Twitter object will be instantiated and called as such:
 * Twitter* obj = new Twitter();
 * obj->postTweet(userId,tweetId);
 * vector<int> param_2 = obj->getNewsFeed(userId);
 * obj->follow(followerId,followeeId);
 * obj->unfollow(followerId,followeeId);
 */